### PR TITLE
SOV-3036 separate margin trade into module

### PR DESCRIPTION
**Pull request by cwsnt**

### CODE REVIEW

**Review by tjcloa — CHANGES_REQUESTED** [review]
for the sake of clarity and simplicity pls consider the following re-design:  
1) rename LoanTokenLogicTradeLM/WRBTC back to LoanTokenLogicLM/WRBTC because most of the functionality stays there & it is easier to review - compare extracted to another modulle functionality  and do the same same with LoanTokenLogic - just remove extracted functions from the original contract; the extracted mint and burn functions will go to LoanTokenMintAndBurn (see details below)
2) rename LoanTokenLogicTradeLM/WRBTC to LoanTokenMintAndBurn (LoanTokenMintAndBurnWRBTC) which is exactly what it does
2.1.) because LoanTokenMinAndBurn/WRBTC are not shared and the mint & burn functions are isolated, just add the functions code to the contracts so that `getListFunctionSignatures()` and the functions code are at the same contract which simplifies the implementation and improves readability and clarity - put LoanTokenMinAndBurn/WRBTC in beaconLogicLM/WRBTC folders respectively
*you would just need to replace LoanTokenLogicLM with the new contract version (with the mynt & burn removed) and add the new module LoanTokenMintAndBurn (LoanTokenMintAndBurnWRBTC)
3) add the modules replacement onchain tests and/or make sure it is covered by QA
4) adjust the mock contracts and deployment scripts and helpers accordingly

**Review by tjcloa — CHANGES_REQUESTED** [review]
almost there  
pls see inline changes request  - those are just some chore and clarity issues

File: deployment/deploy/2010-deploy-ReplaceLoanTokenModules.js [hunk 1]
```diff
@@ -216,12 +304,18 @@ const func = async function (hre) {
...
+            loanTokenMintAndBurnDeployment.address
+        );
         await loanTokenLogicBeaconWRBTCContract.registerLoanTokenModule(
             loanTokenLogicWRBTCDeployment.address
         );
         await loanTokenLogicBeaconWRBTCContract.registerLoanTokenModule(
             loanTokenSettingsLowerAdminDeployment.address
         );
+        await loanTokenLogicBeaconWRBTCContract.registerLoanTokenModule(
+            loanTokenMintAndBurnWRBTCDeployment.address
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply deployment/deploy/2010-deploy-ReplaceLoanTokenModules.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/515/commits/9c48789ef60e609a9e7fd309b7aad7d62bf9c0fa)

File: deployment/deploy/2010-deploy-ReplaceLoanTokenModules.js [hunk 2]
```diff
@@ -165,7 +209,51 @@ const func = async function (hre) {
             );
         }
 
-        /** 4. Registering Loan Protocol Settings Module to LoanTokenLogicBeaconWrbtc */
+        /** 5. Registering Loan Token Mint and Burn Wrbtc Module to LoanTokenLogicBeaconWrbtc */
+        activeModuleIndex = await loanTokenLogicBeaconWRBTCContract.activeModuleIndex(
+            ethers.utils.formatBytes32String(modulesList.LoanTokenMintAndBurnWrbtc)
+        );
+        moduleImplAddress = await loanTokenLogicBeaconWRBTCContract.moduleUpgradeLog(
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply deployment/deploy/2010-deploy-ReplaceLoanTokenModules.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/515/commits/9c48789ef60e609a9e7fd309b7aad7d62bf9c0fa)

File: deployment/deploy/2010-deploy-ReplaceLoanTokenModules.js [hunk 3]
```diff
@@ -165,7 +209,51 @@ const func = async function (hre) {
             );
         }
 
-        /** 4. Registering Loan Protocol Settings Module to LoanTokenLogicBeaconWrbtc */
+        /** 5. Registering Loan Token Mint and Burn Wrbtc Module to LoanTokenLogicBeaconWrbtc */
+        activeModuleIndex = await loanTokenLogicBeaconWRBTCContract.activeModuleIndex(
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply deployment/deploy/2010-deploy-ReplaceLoanTokenModules.js]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/515/commits/9c48789ef60e609a9e7fd309b7aad7d62bf9c0fa)

File: contracts/connectors/loantoken/LoanTokenLogicTrade.sol [hunk 4]
```diff
@@ -0,0 +1,1004 @@
...
+ * Copyright 2017-2021, bZeroX, LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0.
+ */
+
+pragma solidity 0.5.17;
+pragma experimental ABIEncoderV2;
+
+import "./LoanTokenLogicStandard.sol";
+
+contract LoanTokenLogicTrade is LoanTokenLogicStandard {
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply contracts/connectors/loantoken/LoanTokenLogicTrade.sol]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/515/commits/9c48789ef60e609a9e7fd309b7aad7d62bf9c0fa)

File: contracts/connectors/loantoken/modules/beaconLogicLM/LoanTokenLogicLM.sol [hunk 5]
```diff
@@ -55,55 +60,15 @@ contract LoanTokenLogicLM is LoanTokenLogicStandard {
          * LoanTokenLogicStandard also has mint & burn function (overloading).
```
[the remark this answers is not in the document]
Comment (cwsnt) [reply contracts/connectors/loantoken/modules/beaconLogicLM/LoanTokenLogicLM.sol]: updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/515/commits/9c48789ef60e609a9e7fd309b7aad7d62bf9c0fa)

**Review by ororopickpocket — COMMENTED** [review]
Looks good in general, but i do not understand why you decided to split the LoanTokenLogicStandard into LoanTokenLogicStandard and LoanTokenLogicSplit if one is inheriting from the other. If I'm not mistaken, inheritance does not affect the contract size, so you could just as well leave it in one piece. Care to explain your reasoning? Am I wrong?

**Review by tjcloa — CHANGES_REQUESTED** [review]
consider the following changes:  
1) rename LoanTokenLogicLM to LoanTokenLogic
2) rename LoanTokenMyntAndBurn to LoanTokenLogicLM and inherit from LoanTokenLogicSplit
3) rename LoanTokenMintAndBurnWrbtc to LoanTokenLogicWrbtcLM and inherit from LoanTokenLogicSplit
4) LoanTokenLogicStandard should not inherit mint and burn from LoanTokenLogicSplit because it doesn't use them. Extract internal shared by the modules functions into a separate contract - this way it will be much cleaner and will release even more space.

File: contracts/connectors/loantoken/modules/beaconLogicWRBTC/LoanTokenMintAndBurnWrbtc.sol [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ * Copyright 2017-2020, bZeroX, LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0.
+ */
+
+pragma solidity 0.5.17;
+pragma experimental ABIEncoderV2;
+
+import "../../LoanTokenLogicStandard.sol";
+
+contract LoanTokenMintAndBurnWrbtc is LoanTokenLogicStandard {
```
Comment (ororopickpocket) [inline contracts/connectors/loantoken/modules/beaconLogicWRBTC/LoanTokenMintAndBurnWrbtc.sol:11]: Why are you here not also inheriting from LoanTokenLogicSplit ?

Comment (cwsnt) [reply contracts/connectors/loantoken/modules/beaconLogicWRBTC/LoanTokenMintAndBurnWrbtc.sol]: Updated in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/515/commits/7fee23c33c9468d4e032cd3d628da2a5a61a1138)

**Review by tjcloa — APPROVED** [review]
lgtm
approved, but left 1 natspec comment

**Review by ororopickpocket — APPROVED** [review]

**Pull request merged**